go/marshal/encode.go

```diff
@@ -223,18 +260,29 @@ func structEncoder(t reflect.Type, parentStructTypes []reflect.Type) encoderFunc
 
 	parentStructTypes = append(parentStructTypes, t)
 	fields, structType, originalFieldIndex := typeFields(t, parentStructTypes)
-	if structType != nil {
+
+	// It won't be possible to use the pre-computed struct type if it implements
+	// MarshalOptionsGetter, because the struct name can change at runtime.
+	isOptsGetter := t.Implements(optionsGetterType)
+
+	if originalFieldIndex != nil {
+		// Slowest path - we are extending some other struct. We need to start with the
+		// type of that struct and extend.
 		e = func(v reflect.Value) types.Value {
-			values := make([]types.Value, len(fields))
-			for i, f := range fields {
-				values[i] = f.encoder(v.Field(f.index))
+			ret := v.FieldByIndex(originalFieldIndex).Interface().(types.Struct)
+			for _, f := range fields {
+				fv := v.Field(f.index)
+				if !fv.IsValid() || f.omitEmpty && isEmptyValue(fv) {
+					continue
+				}
+				ret = ret.Set(f.name, f.encoder(fv))
 			}
-			return types.NewStructWithType(structType, values)
+			return ret
 		}
-	} else if originalFieldIndex == nil {
-		// Slower path: cannot precompute the Noms type since there are Noms collections,
-		// but at least there are a set number of fields
-		name := strings.Title(t.Name())
+	} else if structType == nil || isOptsGetter {
+		// Slower path: cannot precompute the Noms type since there are Noms
+		// collections, or behavior might change at marshal-time via
+		// MarshalOptions, but at least there are a set number of fields.
```

Comment:
Another option might be to cache this based on name too since I don't think it is common to change the name at runtime.

Maybe a TODO is sufficient since we do not yet know if this will be a performance critical.